Reporting API - tidy up CSPViolationReportBody (#35602)
mdn/content: files/en-us/web/api/cspviolationreportbody/columnnumber/index.md

slug: Web/API/CSPViolationReportBody/columnNumber
page-type: web-api-instance-property
browser-compat: api.CSPViolationReportBody.columnNumber
---

{{APIRef("Reporting API")}}

The **`columnNumber`** read-only property of the {{domxref("CSPViolationReportBody")}} interface indicates the column number in the source file that triggered the [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/CSP) violation.

Note that the browser extracts the value from _the global object_ of the file that triggered the violation.
If the resource that triggers the CSP violation is not loaded, the value will be `null`.
See {{domxref("CSPViolationReportBody.sourceFile")}} for more information.

This property is most useful alongside {{domxref("CSPViolationReportBody.sourceFile")}} and {{domxref("CSPViolationReportBody.lineNumber")}}, as it provides the location of the column in that file and line that resulted in a violation.

## Value

An integer containing the column number that triggered the violation, or `null`.

## Examples

### CSP inline script violation

This example triggers a CSP violation using an inline script, and reports the violation using a {{domxref("ReportingObserver")}}.

#### HTML

> The HTML file below uses the [`<meta>`](/en-US/docs/Web/HTML/Element/meta) element to set the {{httpheader('Content-Security-Policy')}} `default-src` to `self`, which allows scripts and other resources to be loaded from the same domain, but does not allow inline scripts to be executed.

Reviewer: ```suggestion
The HTML file below uses the [`<meta>`](/en-US/docs/Web/HTML/Element/meta) element to set the {{httpheader('Content-Security-Policy')}} `default-src` to `self`, which allows scripts and other resources to be loaded from the same origin, but does not allow inline scripts to be executed.
```

</html>
```

#### JavaScript (main.js)

The document above also loads the external script `main.js`, which is shown below.
Because this is loaded from the same domain as the HTML, it is not blocked by the CSP.

The script creates a new {{domxref("ReportingObserver")}} to observe content violation reports of type `"csp-violation"`.
Each time the callback function is invoked, we get the body of the first entry of the reports array, and use it to log the file, line, and column of the violation to the console.

```js
// main.js
const observer = new ReportingObserver(
  (reports, observer) => {
    const cspViolationBody = reports[0].body;
    console.log(`sourceFile: ${cspViolationBody.sourceFile}`);
    console.log(`lineNumber: ${cspViolationBody.lineNumber}`);
    console.log(`columnNumber: ${cspViolationBody.columnNumber}`);
  },
  {
    types: ["csp-violation"],
    buffered: true,
  },
);

observer.observe();
```

Note that while there might be multiple reports in the returned array, for brevity we only log the values of the first element.

#### Results

If serving the above code using a local server (on `http://127.0.0.1:9999/`), the output of the log on Chrome is:

```plain

> sourceFile: http://127.0.0.1:9999/inline/

Author: ```suggestion
sourceFile: http://127.0.0.1:9999/test/
```

lineNumber: 17
columnNumber: 0
```

Note that the column number is (incorrectly) set to `0`.
The result is similar for Firefox, which correctly reports the `columnNumber`.

```plain

> sourceFile: http://127.0.0.1:9999/inline/

Author: ```suggestion
sourceFile: http://127.0.0.1:9999/test/
```

lineNumber: 17

> columnNumber: 13

Author: Testing shows that  if you have a line like `<script>ab`, the value would be the `9` (i.e. 8 chars to just after `>` and then one more. It makes no particular sense, but I have added this information anyway.

> columnNumber: 13

Author: @wbamberg Further, I prettified this, then ran the code exactly "as-is" to test, and got the values above. This is the best we can do to make sure that others get the same results.

    console.log(`lineNumber: ${cspViolationBody.lineNumber}`);
    console.log(`columnNumber: ${cspViolationBody.columnNumber}`);
  },
  {
    types: ["csp-violation"],
    buffered: true,
  },
);

observer.observe();
```

Note that while there might be multiple reports in the returned array, for brevity we only log the values of the first element.

#### Results

You can try this out using a [local server](/en-US/docs/Learn/Common_questions/Tools_and_setup/set_up_a_local_testing_server).
Copy the above code into `test/index.html` and `test/main.js` and run the server in the root directory.
Assuming the address of the local server is `http://127.0.0.1:9999`, you can then load the HTML file from `http://127.0.0.1:9999/test/` (or `http://127.0.0.1:9999/test/index.html`).

With the above setup, the output of the log on Chrome is:

```plain
sourceFile: http://127.0.0.1:9999/test/
lineNumber: 15
columnNumber: 0
```

The result is similar for Firefox:

```plain
sourceFile: http://127.0.0.1:9999/test/
lineNumber: 15
columnNumber: 13
```

Note that the column number is different for the two browsers.
Chrome always appears to report `0`.

> The value on Firefox represents the first character after the end of the opening `<script>` element.

Reviewer: ```suggestion
The value on Firefox represents the position of the first character after the end of the opening `<script>` element.
```

> The value on Firefox appears to point to the second character after the end of the opening `<script>` element.

Reviewer: ```suggestion
The value on Firefox represents the first character after the end of the opening `<script>` element.
```

I think it's first, not second, because it's not zero-based. Compare say with VSCode:

<img width="248" alt="Screen Shot 2024-09-20 at 9 47 30 PM" src="https://github.com/user-attachments/assets/04e0ff28-836b-4568-a0f9-c7fa4b8f1d6b">

---
title: "CSPViolationReportBody: columnNumber property"
short-title: columnNumber
slug: Web/API/CSPViolationReportBody/columnNumber
page-type: web-api-instance-property
browser-compat: api.CSPViolationReportBody.columnNumber
---

{{APIRef("Reporting API")}}

The **`columnNumber`** read-only property of the {{domxref("CSPViolationReportBody")}} interface indicates the column in the source file that caused the reported CSP violation.

This property is most useful alongside {{domxref("CSPViolationReportBody.sourceFile")}} and {{domxref("CSPViolationReportBody.lineNumber")}}, as it provides the location of the column in that file and line where the error occurred.

## Value

An integer, or `null` if the column is not known.

> ## Examples

Author: @wbamberg You're right of course. I had originally written test code to show this but could only ever get `null` for these values. 

I have updated the description and provided an example. The behaviour is a bit "unexpected".

> ## Examples

Author: I'll do the others too, but this one could be looked at now, as it sets the pattern.
